test/socket.js

```diff
@@ -311,3 +312,286 @@ test('try send after close', async function (t) {
 
   t.is(a.trySend(Buffer.from('hello'), a.address().port), undefined)
 })
+
+test('connect to invalid host ip', function (t) {
+  t.plan(1)
+
+  const u = new UDX()
+
+  const a = u.createSocket()
+  const s = u.createStream(1)
+
+  const invalidHost = '0.-1.0.0'
+
+  try {
+    s.connect(a, 2, 0, invalidHost)
+  } catch (error) {
+    t.is(error.message, `${invalidHost} is not a valid IP address`)
+  }
+})
+
+test('bind to invalid host ip', function (t) {
+  t.plan(1)
+
+  const u = new UDX()
+  const a = u.createSocket()
+
+  const invalidHost = '0.-1.0.0'
+
+  try {
+    a.bind(0, invalidHost)
+  } catch (error) {
+    t.is(error.message, `${invalidHost} is not a valid IP address`)
+  }
+})
+
+test('send to invalid host ip', async function (t) {
+  t.plan(1)
+
+  const u = new UDX()
+  const a = u.createSocket()
+
+  a.bind(0)
+
+  const invalidHost = '0.-1.0.0'
+
+  try {
+    await a.send(Buffer.from('hello'), a.address().port, invalidHost)
+  } catch (error) {
+    t.is(error.message, `${invalidHost} is not a valid IP address`)
+  }
+
+  await a.close()
+})
+
+test('try send to invalid host ip', async function (t) {
+  t.plan(1)
+
+  const u = new UDX()
+  const a = u.createSocket()
+
+  a.bind(0)
+
+  const invalidHost = '0.-1.0.0'
+
+  try {
+    a.trySend(Buffer.from('hello'), a.address().port, invalidHost)
+  } catch (error) {
+    t.is(error.message, `${invalidHost} is not a valid IP address`)
+  }
+
+  await a.close()
+})
+
+test('send without bind', async function (t) {
+  t.plan(1)
+
+  const u = new UDX()
+
+  const a = u.createSocket()
+  const b = u.createSocket()
+
+  b.on('message', function (message) {
+    console.log('recv')
+    t.alike(message, Buffer.from('hello'))
+    a.close()
+    b.close()
+  })
+
+  b.bind(0)
+  await a.send(Buffer.from('hello'), b.address().port)
+})
+
+test('try send without bind', async function (t) {
+  t.plan(1)
+
+  const u = new UDX()
+
+  const a = u.createSocket()
+  const b = u.createSocket()
+
+  b.on('message', function (message) {
+    console.log('recv')
+    t.alike(message, Buffer.from('hello'))
+    a.close()
+    b.close()
+  })
+
+  b.bind(0)
+  a.trySend(Buffer.from('hello'), b.address().port)
+})
+
+test('get address without bind', async function (t) {
+  const u = new UDX()
+  const a = u.createSocket()
+  t.is(a.address(), null)
```

Comment:
close the socket here also after

Reply:
`bind()` sets `this._inited`, and the test doesn't bind, so the `close` event would never fire:
https://github.com/hyperswarm/libudx/blob/f01f56c236f3ec31ae4981c2e49d551171d1c650/lib/socket.js#L149
So if a socket didn't bind then I think there is no need to close